In [34]:
import pandas as pd
import matplotlib.pyplot as plt
plt.rc("font", family="Malgun Gothic")

In [2]:
df1 = pd.read_csv("dataset/01_Contract_Data.csv")

In [3]:
df1

Index  Member_ID Sales_Type Contract_Type Channel    Datetime  Term  \
0          1   66758234         렌탈          일반계약    영업방판  2019-05-06    60   
1          2   66755948         렌탈          교체계약    영업방판  2020-02-20    60   
2          3   66756657         렌탈          일반계약  홈쇼핑/방송  2019-02-28    60   
3          4   66423450        멤버십        멤버십3유형     재계약  2019-05-13    12   
4          5   66423204        멤버십        멤버십3유형     재계약  2019-05-10    12   
...      ...        ...        ...           ...     ...         ...   ...   
51296  51298   66579515         렌탈        프로모션계약   대형마트A  2019-03-01    60   
51297  51299   66799558         렌탈          일반계약   대형마트A  2019-04-01    60   
51298  51300   66799197         렌탈        프로모션계약    영업방판  2019-04-01    39   
51299  51301   66792778         렌탈          일반계약  홈쇼핑/방송  2020-02-06    60   
51300  51302   66799607         렌탈          일반계약  홈쇼핑/방송  2019-04-24    60   

      Payment_Type Product_Type  Amount_Month Customer_Type   Age Address1  \
0              CMS        DES-1         96900            개인  42.0      경기도   
1             카드이체        DES-1        102900            개인  39.0      경기도   
2              CMS        DES-1         96900            개인  48.0      경기도   
3              CMS        DES-1         66900            개인  39.0      경기도   
4              CMS        DES-1         66900            개인  60.0      경기도   
...            ...          ...           ...           ...   ...      ...   
51296          CMS       DES-3A         96900            개인  47.0      경기도   
51297          CMS        DES-1         96900            개인  42.0      경기도   
51298         카드이체          ERA        120900            개인  65.0    서울특별시   
51299         카드이체        DES-1         96900            개인  54.0    서울특별시   
51300          CMS        DES-1         96900            개인  53.0    서울특별시   

      Address2 State  Overdue_count Overdue_Type Gender  Credit_Rank    Bank  
0          경기도  계약확정              0           없음     여자          9.0   새마을금고  
1          경기도  계약확정              0           없음     남자          2.0    현대카드  
2          경기도  계약확정              0           없음     여자          8.0    우리은행  
3          경기도  계약확정              0           없음     남자          5.0  농협회원조합  
4          경기도  기간만료             12           있음     남자          8.0  농협회원조합  
...        ...   ...            ...          ...    ...          ...     ...  
51296      경기도  계약확정              0           없음     남자          NaN    기업은행  
51297      경기도  계약확정              0           없음     여자          8.0   새마을금고  
51298    서울특별시  계약확정              0           없음     여자          1.0    롯데카드  
51299    서울특별시  계약확정              0           없음     여자          2.0    롯데카드  
51300    서울특별시  계약확정              0           없음     여자          8.0    신한은행  

[51301 rows x 20 columns]

In [4]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51301 entries, 0 to 51300
Data columns (total 20 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Index          51301 non-null  int64  
 1   Member_ID      51301 non-null  int64  
 2   Sales_Type     51301 non-null  object 
 3   Contract_Type  51301 non-null  object 
 4   Channel        51301 non-null  object 
 5   Datetime       51301 non-null  object 
 6   Term           51301 non-null  int64  
 7   Payment_Type   51301 non-null  object 
 8   Product_Type   51301 non-null  object 
 9   Amount_Month   51301 non-null  int64  
 10  Customer_Type  51299 non-null  object 
 11  Age            44329 non-null  float64
 12  Address1       51299 non-null  object 
 13  Address2       51299 non-null  object 
 14  State          51301 non-null  object 
 15  Overdue_count  51301 non-null  int64  
 16  Overdue_Type   51301 non-null  object 
 17  Gender         51301 non-null  object 
 18  Credit

In [6]:
# 1. Channel별 Amount_Month의 mean
#df1["Channel"].value_counts()
df1.pivot_table(index="Channel", values="Amount_Month", aggfunc="mean")

Amount_Month
Channel               
대형마트A     97016.330159
대형마트C     95276.974790
대형마트E    100423.647295
대형마트H    100025.333333
대형마트N     76514.457831
영업방판     100322.794631
자체홈페이지    98402.500000
재계약       67176.744186
전단홍보     104048.571429
전문매장H     72514.008677
전문매장Z     98127.363184
직영계열사A   104970.000000
직영계열사B   111300.000000
직영유통사    119011.117021
홈쇼핑/방송    89092.232411
홈쇼핑/인터넷   96685.600000

In [9]:
# 2. Channel과 Gender별 Amount_Month의 mean
df1.pivot_table(index=["Channel", "Gender"], values="Amount_Month", aggfunc="mean")

Amount_Month
Channel Gender               
대형마트A   남자       96521.515368
        여자       97224.461816
대형마트C   여자       95276.974790
대형마트E   남자      100724.414062
        여자      100267.309645
대형마트H   남자      100611.111111
        여자       99695.833333
대형마트N   남자       76119.512195
        여자       76900.000000
영업방판    남자      100059.535067
        여자      100446.886646
자체홈페이지  남자      100281.250000
        여자       97150.000000
재계약     남자       67257.421875
        여자       67133.263158
전단홍보    남자      105357.485030
        여자      103492.366412
전문매장H   남자       99922.424242
        여자       67547.989019
전문매장Z   남자       97738.163265
        여자       98393.715084
직영계열사A  여자      104970.000000
직영계열사B  여자      111300.000000
직영유통사   여자      119011.117021
홈쇼핑/방송  남자       88578.558875
        여자       89319.713142
홈쇼핑/인터넷 남자       96781.188119
        여자       96650.364964

In [10]:
# 3. Gender와 Channel에 따른 Amount_Month의 sum
df1.pivot_table(index=["Gender", "Channel"], values="Amount_Month", aggfunc="sum")

Amount_Month
Gender Channel              
남자     대형마트A       135033600
       대형마트E        51570900
       대형마트H         8149500
       대형마트N         3120900
       영업방판        761853300
       자체홈페이지        4813500
       재계약          51653700
       전단홍보         17594700
       전문매장H        49461600
       전문매장Z        47891700
       홈쇼핑/방송      352808400
       홈쇼핑/인터넷       9774900
여자     대형마트A       323368560
       대형마트C        11337960
       대형마트E        98763300
       대형마트H        14356200
       대형마트N         3229800
       영업방판       1622518560
       자체홈페이지        6994800
       재계약          95664900
       전단홍보         40672500
       전문매장H       184541106
       전문매장Z        70449900
       직영계열사A        2099400
       직영계열사B        2114700
       직영유통사        22374090
       홈쇼핑/방송      803341500
       홈쇼핑/인터넷      26482200

In [12]:
# 4. Gender와 Channel에 따른 Amount_Month와 Overdue_count의 sum
df1.pivot_table(index=["Gender", "Channel"], values=["Amount_Month", "Overdue_count"], aggfunc="sum")

Amount_Month  Overdue_count
Gender Channel                             
남자     대형마트A       135033600            351
       대형마트E        51570900             89
       대형마트H         8149500             21
       대형마트N         3120900              0
       영업방판        761853300           2083
       자체홈페이지        4813500              1
       재계약          51653700            167
       전단홍보         17594700             21
       전문매장H        49461600            128
       전문매장Z        47891700             43
       홈쇼핑/방송      352808400            476
       홈쇼핑/인터넷       9774900             20
여자     대형마트A       323368560            553
       대형마트C        11337960              0
       대형마트E        98763300            103
       대형마트H        14356200             22
       대형마트N         3229800              0
       영업방판       1622518560           2804
       자체홈페이지        6994800              3
       재계약          95664900            148
       전단홍보         40672500             80
       전문매장H       184541106             64
       전문매장Z        70449900             88
       직영계열사A        2099400              0
       직영계열사B        2114700              0
       직영유통사        22374090            172
       홈쇼핑/방송      803341500            795
       홈쇼핑/인터넷      26482200             47

In [13]:
# 5. Gender와 Channel에 따른 Amount_Month와 Overdue_count의 sum, min, max
df1.pivot_table(index=["Gender", "Channel"], 
                values=["Amount_Month", "Overdue_count"],
               aggfunc=["sum", "min", "max"])

sum                        min                \
               Amount_Month Overdue_count Amount_Month Overdue_count   
Gender Channel                                                         
남자     대형마트A      135033600           351        55200             0   
       대형마트E       51570900            89        81900             0   
       대형마트H        8149500            21        87900             0   
       대형마트N        3120900             0        75900             0   
       영업방판       761853300          2083        78900             0   
       자체홈페이지       4813500             1        84900             0   
       재계약         51653700           167        55200             0   
       전단홍보        17594700            21        87900             0   
       전문매장H       49461600           128        81900             0   
       전문매장Z       47891700            43        81900             0   
       홈쇼핑/방송     352808400           476        80400             0   
       홈쇼핑/인터넷      9774900            20        80400             0   
여자     대형마트A      323368560           553        55200             0   
       대형마트C       11337960             0        87930             0   
       대형마트E       98763300           103        81900             0   
       대형마트H       14356200            22        81900             0   
       대형마트N        3229800             0        75900             0   
       영업방판      1622518560          2804        78900             0   
       자체홈페이지       6994800             3        78900             0   
       재계약         95664900           148        55200             0   
       전단홍보        40672500            80        66900             0   
       전문매장H      184541106            64        54603             0   
       전문매장Z       70449900            88        81900             0   
       직영계열사A       2099400             0        82200             0   
       직영계열사B       2114700             0        93900             0   
       직영유통사       22374090           172        96900             0   
       홈쇼핑/방송     803341500           795        72900             0   
       홈쇼핑/인터넷     26482200            47        80400             0   

                        max                
               Amount_Month Overdue_count  
Gender Channel                             
남자     대형마트A         185700            12  
       대형마트E         185700            13  
       대형마트H         142200            12  
       대형마트N          78900             0  
       영업방판          215700            15  
       자체홈페이지        126900             1  
       재계약           105900            12  
       전단홍보          185700             7  
       전문매장H         145200            13  
       전문매장Z         145200            10  
       홈쇼핑/방송        145200            15  
       홈쇼핑/인터넷       138900            13  
여자     대형마트A         185700            12  
       대형마트C         126900             0  
       대형마트E         185700            14  
       대형마트H         144900            12  
       대형마트N         102900             0  
       영업방판          215700            15  
       자체홈페이지        105900             2  
       재계약            96600            12  
       전단홍보          215700            13  
       전문매장H         145200            14  
       전문매장Z         145200            13  
       직영계열사A        112200             0  
       직영계열사B        126900             0  
       직영유통사         180750             1  
       홈쇼핑/방송        144900            14  
       홈쇼핑/인터넷       124200             8

In [15]:
# 6. Sales_Type에 따른 Datetime별 Amount_Month의 sum
df1.pivot_table(index=["Sales_Type"], columns="Datetime", values="Amount_Month", aggfunc="sum")

Datetime    2019-01-01  2019-01-02  2019-01-03  2019-01-04  2019-01-05  \
Sales_Type                                                               
렌탈           4346700.0   4957800.0   7011300.0   5276700.0   5170950.0   
멤버십           209700.0     66900.0    255900.0    133800.0    334500.0   

Datetime    2019-01-06  2019-01-07  2019-01-08  2019-01-09  2019-01-10  ...  \
Sales_Type                                                              ...   
렌탈           4445700.0   8282400.0   5970150.0   5577600.0   6996900.0  ...   
멤버십           142800.0    255900.0         NaN    133800.0    133800.0  ...   

Datetime    2020-07-29  2020-07-30  2020-07-31  2020-08-01  2020-08-02  \
Sales_Type                                                               
렌탈           5478600.0   7551300.0  10361100.0   6187500.0   2779200.0   
멤버십           302361.0    383322.0    730365.0    671079.0    188403.0   

Datetime    2020-08-03  2020-08-04  2020-08-05  2020-08-06  2020-08-07  
Sales_Type                                                              
렌탈           2953500.0   2732700.0   3034260.0   2523000.0   2111100.0  
멤버십           192039.0    125643.0    245382.0    247758.0     58743.0  

[2 rows x 577 columns]

In [23]:
df1["Datetime(dt)"] = pd.to_datetime(df1["Datetime"], format="%Y-%m-%d")
df1["Year"] = df1["Datetime(dt)"].dt.year
df1["Month"] = df1["Datetime(dt)"].dt.month

In [24]:
# 7. Channel에 따른 월별  Amount_Month의 sum
df1.pivot_table(index="Channel", columns="Month", 
                values="Amount_Month",
               aggfunc="sum")

Month             1            2            3            4            5   \
Channel                                                                    
대형마트A     20271300.0   96975840.0   54780510.0   42882150.0   36463050.0   
대형마트C       694380.0    3163260.0    1158000.0     372390.0     768030.0   
대형마트E     20201100.0   25190400.0   18663000.0   13513200.0   11181300.0   
대형마트H      3527400.0    3105600.0    2656800.0    2491500.0    1734600.0   
대형마트N       531300.0    1068600.0     837900.0     683100.0     315600.0   
영업방판     316174200.0  393566520.0  304034250.0  232733160.0  180750690.0   
자체홈페이지      582900.0    2437500.0    1681800.0     802200.0     881100.0   
재계약       20934300.0   22681800.0   18390900.0   12398400.0   11508300.0   
전단홍보       7683000.0   10033500.0    6838500.0    5909700.0    4656300.0   
전문매장H     23965248.0   40271526.0   31000287.0   23404260.0   19175844.0   
전문매장Z     15642000.0   18392100.0   15526200.0   13044600.0    9744900.0   
직영계열사A           NaN     322800.0     101400.0     112200.0          NaN   
직영계열사B       93900.0     187800.0     376800.0      93900.0     123000.0   
직영유통사      2324280.0    4172880.0    3018900.0    1844160.0    2163810.0   
홈쇼핑/방송   154646700.0  181100700.0  145563300.0  112216200.0   89219100.0   
홈쇼핑/인터넷    5244000.0    5890500.0    3345600.0    4594500.0    3840000.0   

Month             6            7            8            9            10  \
Channel                                                                    
대형마트A     34781700.0   47744580.0   25986870.0   21650280.0   31277250.0   
대형마트C       858150.0    1758930.0     736170.0     117900.0    1053240.0   
대형마트E     10683900.0   13498500.0    7335600.0    6993300.0    9118800.0   
대형마트H      1180800.0    2055000.0    1386600.0     923100.0    1550400.0   
대형마트N       455400.0     607200.0     151800.0     610200.0     607200.0   
영업방판     159866280.0  226868760.0  123142230.0  108835530.0  142402860.0   
자체홈페이지     1058400.0    1003800.0     579900.0     480000.0    1425600.0   
재계약       11766600.0   15115200.0    8458500.0    6819000.0    7228200.0   
전단홍보       4391700.0    5724000.0    2408100.0    2714400.0    3603300.0   
전문매장H     16514739.0   23900502.0   10093590.0   11433087.0   14434515.0   
전문매장Z      7788000.0   10469400.0    7516500.0    5085300.0    7008300.0   
직영계열사A           NaN     112200.0     224400.0     224400.0     306600.0   
직영계열사B      253800.0     117900.0     123000.0     187800.0     246000.0   
직영유통사      1848180.0    1064100.0    1990410.0    1287750.0    1399710.0   
홈쇼핑/방송    82573800.0  112373100.0   56945400.0   52394100.0   74601000.0   
홈쇼핑/인터넷    2745600.0    2784600.0    1638300.0    2052900.0    1634700.0   

Month            11           12  
Channel                           
대형마트A    15115500.0   30473130.0  
대형마트C       87930.0     569580.0  
대형마트E     4553700.0    9401400.0  
대형마트H     1020300.0     873600.0  
대형마트N      151800.0     330600.0  
영업방판     70206300.0  125791080.0  
자체홈페이지     495000.0     380100.0  
재계약       3965700.0    8051700.0  
전단홍보      1487400.0    2817300.0  
전문매장H     6533370.0   13275738.0  
전문매장Z     2485800.0    5638500.0  
직영계열사A     306600.0     388800.0  
직영계열사B     216900.0      93900.0  
직영유통사      246000.0    1013910.0  
홈쇼핑/방송   34065000.0   60451500.0  
홈쇼핑/인터넷   1268700.0    1217700.0

In [25]:
# 8. Sales_Type에 따른 월별 Amount_Month의 mean
df1.pivot_table(index="Sales_Type",
                columns="Month", 
               values="Amount_Month",
               aggfunc="sum")

Month              1          2          3          4          5          6   \
Sales_Type                                                                     
렌탈          564577560  754111500  569920260  438592560  348396180  313715910   
멤버십          27938448   54449826   38053887   28503060   24129444   23051139   

Month              7          8          9          10         11         12  
Sales_Type                                                                    
렌탈          433002870  232279680  206670360  280292160  133427430  243143700  
멤버십          32194902   16437690   15138687   17605515    8778570   17624838

In [33]:
# 9. Sales_Type에 따른 월별 Amount_Month의 sum (월별 매출의 추이)
import matplotlib.pyplot as plt
plt.rc("font", family="Malgun Gothic")
df1.pivot_table(index="Sales_Type", columns="Month", values="Amount_Month", aggfunc="sum")#.transpose().plot(subplots=True)

Month              1          2          3          4          5          6   \
Sales_Type                                                                     
렌탈          564577560  754111500  569920260  438592560  348396180  313715910   
멤버십          27938448   54449826   38053887   28503060   24129444   23051139   

Month              7          8          9          10         11         12  
Sales_Type                                                                    
렌탈          433002870  232279680  206670360  280292160  133427430  243143700  
멤버십          32194902   16437690   15138687   17605515    8778570   17624838

In [35]:
# 10. Sales_Type에 따른 연도별, 월별 Amount_Month의 sum
df1.pivot_table(index="Sales_Type", columns=["Year", "Month"], values="Amount_Month", aggfunc="sum")

Year             2019                                                         \
Month              1          2          3          4          5          6    
Sales_Type                                                                     
렌탈          394690080  581147940  423192960  318915930  263786940  236969310   
멤버십          17205900   44238291   27736254   20259312   19053555   17510016   

Year                                                                          \
Month              7          8          9          10         11         12   
Sales_Type                                                                     
렌탈          313217310  209958420  206670360  280292160  133427430  243143700   
멤버십          24422919   14708643   15138687   17605515    8778570   17624838   

Year             2020                                                       \
Month              1          2          3          4         5         6    
Sales_Type                                                                   
렌탈          169887480  172963560  146727300  119676630  84609240  76746600   
멤버십          10732548   10211535   10317633    8243748   5075889   5541123   

Year                             
Month              7         8   
Sales_Type                       
렌탈          119785560  22321260  
멤버십           7771983   1729047

In [38]:
# 11. Bank에 따른 월별 Amount_Month의 sum
#df1["Bank"].value_counts()
df1.pivot_table(index="Bank", columns="Month", values="Amount_Month", aggfunc="sum")

Month               1            2            3           4           5   \
Bank                                                                       
BC카드        13861050.0   17040600.0   18356100.0  13517400.0   8998500.0   
NH농협카드        105900.0          NaN          NaN         NaN    126900.0   
SC제일은행       5866650.0    6409050.0    5074800.0   4279650.0   3644250.0   
SK증권               NaN          NaN          NaN         NaN         NaN   
경남은행         5560200.0    6782100.0    6406800.0   4664100.0   3291000.0   
광주은행         4374600.0    6754500.0    4083900.0   3750000.0   1994400.0   
국민은행        70656600.0   91146900.0   74246100.0  55984200.0  43775679.0   
국민카드        16005300.0   19740600.0   17868900.0  12062700.0   8166300.0   
기업은행        26801100.0   33640350.0   21714450.0  16829250.0  14071800.0   
농협중앙회       44864700.0   61123800.0   48209700.0  37406400.0  29891100.0   
농협회원조합      74546979.0  100832100.0   73452000.0  57693600.0  50416800.0   
대구은행         8161500.0   12276000.0    7927500.0   8072400.0   5581200.0   
대신증권               NaN     105900.0          NaN         NaN         NaN   
대우증권               NaN      81900.0          NaN         NaN         NaN   
롯데카드       118315800.0  143077200.0  111853500.0  83440500.0  69893100.0   
메리츠종합금융증권          NaN          NaN          NaN         NaN         NaN   
미래에셋증권         96900.0     102900.0     105900.0         NaN         NaN   
부산은행        10291500.0   16880400.0   12106800.0   8553600.0   5594400.0   
산림조합중앙회            NaN          NaN          NaN         NaN         NaN   
산업은행          193800.0     357000.0          NaN    208800.0    193800.0   
삼성증권           96900.0          NaN      66900.0    172800.0         NaN   
삼성카드         9261000.0   14161200.0   10396800.0   8596200.0   5322900.0   
상호저축은행         96900.0          NaN          NaN         NaN         NaN   
새마을금고       12299100.0   16074600.0   12284400.0  10364700.0   6640800.0   
수협중앙회        1671300.0    3303000.0    1872900.0   1846500.0    605700.0   
수협카드               NaN          NaN          NaN         NaN         NaN   
신한금융투자         96900.0      75900.0      96900.0         NaN         NaN   
신한은행        40056300.0   60579750.0   39295500.0  33478500.0  25483050.0   
신한카드        17709000.0   25111800.0   18987000.0  14106600.0  12410100.0   
신협중앙회        4192200.0    6108300.0    4716900.0   1843800.0   2574300.0   
씨티은행         1817100.0    2474700.0    2409900.0   1335600.0   1014900.0   
씨티카드               NaN     124200.0          NaN         NaN         NaN   
외환은행         7853100.0    9239400.0    7356900.0   6084600.0   3836700.0   
외환카드         6999300.0    8925000.0    7494900.0   4365300.0   3291000.0   
우리은행        39301200.0   55235700.0   40988400.0  29855400.0  22415700.0   
우리카드               NaN      96900.0      87900.0         NaN         NaN   
우체국          7668900.0   10329900.0    9039900.0   8025600.0   6688800.0   
유안타증권         163800.0     457500.0     202800.0    284700.0    368100.0   
전북은행         1989000.0    3481800.0    2084100.0   2124900.0   1452000.0   
제주은행          918000.0     484500.0     624300.0    511800.0    372600.0   
하나SK               NaN     208800.0      81900.0    102900.0         NaN   
하나대투증권             NaN          NaN          NaN         NaN         NaN   
하나은행        19416600.0   22635300.0   16111200.0  12266700.0  12036900.0   
하이투자증권             NaN          NaN          NaN         NaN         NaN   
한국투자증권             NaN      66900.0     105900.0         NaN         NaN   
현대증권          420600.0     108900.0          NaN    151800.0    105900.0   
현대카드         9885000.0   14273400.0    8780100.0   7311600.0   7016700.0   

Month              6           7           8           9           10  \
Bank                                                                    
BC카드        7841400.0  11081400.0   6260400.0   5803800.0   7203600.0   
NH농협카드            NaN    126900.0         NaN         Na

In [39]:
# 12. Bank에 따른 월별 Amount_Month의 sum과 그들의 합 (margins=True)
df1.pivot_table(index="Bank", columns="Month", values="Amount_Month", aggfunc="sum", margins=True)

Month                1            2            3            4            5  \
Bank                                                                         
BC카드        13861050.0   17040600.0   18356100.0   13517400.0    8998500.0   
NH농협카드        105900.0          NaN          NaN          NaN     126900.0   
SC제일은행       5866650.0    6409050.0    5074800.0    4279650.0    3644250.0   
SK증권               NaN          NaN          NaN          NaN          NaN   
경남은행         5560200.0    6782100.0    6406800.0    4664100.0    3291000.0   
광주은행         4374600.0    6754500.0    4083900.0    3750000.0    1994400.0   
국민은행        70656600.0   91146900.0   74246100.0   55984200.0   43775679.0   
국민카드        16005300.0   19740600.0   17868900.0   12062700.0    8166300.0   
기업은행        26801100.0   33640350.0   21714450.0   16829250.0   14071800.0   
농협중앙회       44864700.0   61123800.0   48209700.0   37406400.0   29891100.0   
농협회원조합      74546979.0  100832100.0   73452000.0   57693600.0   50416800.0   
대구은행         8161500.0   12276000.0    7927500.0    8072400.0    5581200.0   
대신증권               NaN     105900.0          NaN          NaN          NaN   
대우증권               NaN      81900.0          NaN          NaN          NaN   
롯데카드       118315800.0  143077200.0  111853500.0   83440500.0   69893100.0   
메리츠종합금융증권          NaN          NaN          NaN          NaN          NaN   
미래에셋증권         96900.0     102900.0     105900.0          NaN          NaN   
부산은행        10291500.0   16880400.0   12106800.0    8553600.0    5594400.0   
산림조합중앙회            NaN          NaN          NaN          NaN          NaN   
산업은행          193800.0     357000.0          NaN     208800.0     193800.0   
삼성증권           96900.0          NaN      66900.0     172800.0          NaN   
삼성카드         9261000.0   14161200.0   10396800.0    8596200.0    5322900.0   
상호저축은행         96900.0          NaN          NaN          NaN          NaN   
새마을금고       12299100.0   16074600.0   12284400.0   10364700.0    6640800.0   
수협중앙회        1671300.0    3303000.0    1872900.0    1846500.0     605700.0   
수협카드               NaN          NaN          NaN          NaN          NaN   
신한금융투자         96900.0      75900.0      96900.0          NaN          NaN   
신한은행        40056300.0   60579750.0   39295500.0   33478500.0   25483050.0   
신한카드        17709000.0   25111800.0   18987000.0   14106600.0   12410100.0   
신협중앙회        4192200.0    6108300.0    4716900.0    1843800.0    2574300.0   
씨티은행         1817100.0    2474700.0    2409900.0    1335600.0    1014900.0   
씨티카드               NaN     124200.0          NaN          NaN          NaN   
외환은행         7853100.0    9239400.0    7356900.0    6084600.0    3836700.0   
외환카드         6999300.0    8925000.0    7494900.0    4365300.0    3291000.0   
우리은행        39301200.0   55235700.0   40988400.0   29855400.0   22415700.0   
우리카드               NaN      96900.0      87900.0          NaN          NaN   
우체국          7668900.0   10329900.0    9039900.0    8025600.0    6688800.0   
유안타증권         163800.0     457500.0     202800.0     284700.0     368100.0   
전북은행         1989000.0    3481800.0    2084100.0    2124900.0    1452000.0   
제주은행          918000.0     484500.0     624300.0     511800.0     372600.0   
하나SK               NaN     208800.0      81900.0     102900.0          NaN   
하나대투증권             NaN          NaN          NaN          NaN          NaN   
하나은행        19416600.0   22635300.0   16111200.0   12266700.0   12036900.0   
하이투자증권             NaN          NaN          NaN          NaN          NaN   
한국투자증권             NaN      66900.0     105900.0          NaN          NaN   
현대증권          420600.0     108900.0          NaN     151800.0     105900.0   
현대카드         9885000.0   14273400.0    8780100.0    7311600.0    7016700.0   
All        581614779.0  769908750.0  584491950.0  449292600.0  357275379.0   

Month                6            7            8            9           10  \
Bank                 

In [41]:
# 13. Product_Type별 Amount_Month의 sum과 전체 합 (제품별합, 전체합)
df1.pivot_table(index="Product_Type", values="Amount_Month", aggfunc="sum", margins=True)

Amount_Month
Product_Type              
DES-1           3695115726
DES-2            787657290
DES-3A           197085060
DES-R4            42328500
ERA               77115600
MMC               22734000
All             4822036176

In [45]:
# 14. Product_Type에 따른 연도별 Amount_Month의 sum (제품별합, 전체합) 
df1.pivot_table(index="Product_Type", columns="Year",
                values="Amount_Month", aggfunc="sum")#.to_excel("result2.xlsx")

Year                2019       2020
Product_Type                       
DES-1         2946278610  748837116
DES-2          631877970  155779320
DES-3A         157162560   39922500
DES-R4          33891300    8437200
ERA             62944500   14171100
MMC             17540100    5193900

In [54]:
# 15. Product_Type에 따른 연도별 Amount_Month의 차이 (19년도 -> 20년도 증감액)
df3 = df1.pivot_table(index="Product_Type", columns="Year",
                values="Amount_Month", aggfunc="sum")

df3["증감액"] = df3[2020] - df3[2019]
df3

Year                2019       2020         증감액
Product_Type                                   
DES-1         2946278610  748837116 -2197441494
DES-2          631877970  155779320  -476098650
DES-3A         157162560   39922500  -117240060
DES-R4          33891300    8437200   -25454100
ERA             62944500   14171100   -48773400
MMC             17540100    5193900   -12346200

In [59]:
# 16. 19년도 1-8월 데이터와 20년도 전체 데이터만 추출(filtering)하기
cond1 = (df1["Year"] == 2019)
cond2 = (df1["Month"] <= 8)
cond3 = (df1["Year"] == 2020)
df4 = df1.loc[(cond1 & cond2) | cond3]

### Big Picture
- pivot_tabel(index, columns, values, aggfunc)
> argument에 list를 넣을 수 있다.
- 정확한 제품별 매출 증감액 & 증감률 구하기
- 이번 시간: `pivot_table`로 데이터를 요약 & 정리
- 다음 시간: `melt`로 정리 & 요약된 데이터를 풀기
- 데이터를 뽑았을 때 정리된 형태로 뽑혔다면, 데이터를 풀어헤쳐야 한다. 
> 이 때, `melt`를 사용한다.

01. Channel별 Amount_Month의 mean
02. Channel과 Gender별 Amount_Month의 mean
03. Gender와 Channel에 따른 Amount_Month의 sum
04. Gender와 Channel에 따른 Amount_Month와 Overdue_count의 sum
05. Gender와 Channel에 따른 Amount_Month와 Overdue_count의 sum, min, max 
06. Datetime별 Sales_Type에 따른 Amount_Month의 sum
> 577 columns 너무 많다. DateTimeIndex 처리해서 월별, 연도별 매출액 계산해보자.
07. Channel에 따른 월별 Amount_Month의 sum
08. Sales_Type에 따른 월별 Amount_Month의 mean
09. Sales_Type에 따른 월별 Amount_Month의 sum (19년, 20년 구분 없는 월별 매출의 추이)
10. Sales_Type에 따른 연도별, 월별 Amount_Month의 sum (19년, 20년 구분 없는 월별 매출의 추이)
11. Bank에 따른 월별 Amount_Month의 sum
12. Bank에 따른 월별 Amount_Month의 sum과 월별 합 (margins=True)
13. Product_Type에 따른 Amount_Month의 sum과 전체 합 (제품별합, 전체합)
14. Product_Type에 따른 연도별 Amount_Month의 sum (제품별합, 전체합) 
15. Product_Type에 따른 연도별 Amount_Month의 차이 (19년도 -> 20년도 증감액)
16. 19년도 1-8월 데이터와 20년도 전체 데이터만 추출(filtering)하기
17. 19년도 1-8월, 20년도 전체(1-8월) product type에 따른 연도별 Amount_Month의 sum
18. 1-8월 증감액 & 증감률
> - `증감률 = 증감액 / 전년도 + 이번년도`
> - 실제로 19년 대비 매출이 약 50%정도 대폭 감소했다. 


- index와 values, aggfunc에 list 형태 구조를 집어 넣을 수 있다.
- 동시에 2개의 범주형 항목으로 데이터를 나누어 2개의 숫자형 항목 데이터에 대한 통계량 계산
- aggfunc: default는 `numpy.mean`이다.
- `margins=True`: 총합 계산
> All이 index 마지막에 붙는다. 모든 은행 총합 값 (강력한 기능)
- Datetime: 영업일자, 계약일자 (csv 파일별 description 페이지가 필요함)
- erp(전사적 자원 관리): 데이터가 모여있는 공간
- 실무에서 많이 보는 항목: 전년 대비 매출 증감률, 고객 증감률
- pivot_table을 잘 사용하면, 다양하고 많은 데이터에서 데이터를 항목별로 나눠서, 특정 값에 대한 통계량을 확인하면서, 인사이트를 얻을 수 있다. 
- `df3[2020] - df3[2019]`: dt index가 아니라 col이름
> 제품별 매출 증감액 (음의 값이므로 줄었다.)
- 정확한 제품별 매출 증감액
> 19년도는 1-12월 데이터가 있지만 20년도는 1-8월 데이터만 있으므로 1-8월 증감만 계산해야 정확한 **증감액**과 **증감률**을 계산할 수 있다.
- 19년도, 20년도에 몇월부터 몇월까지의 데이터가 있는지 확인하는 방법:
> - 20년도:`cond1 = (df1["Year"] == 2020) df1.loc[cond1]["Month"].unique() # 20년도는 정확히 1-8월 데이터만 있음`
> - 19년도: `cond1 = (df1["Year"] == 2019) df1.loc[cond1]["Month"].unique() # 19년도는 정확히 1-12월 데이터가 있음`
- 19년도 1-8월 데이터와 20년도 전체 데이터만 추출(filtering)하기
> - `cond1 = df1["Year"] == 2019 cond2 = (df1["Month"] <= 8) cond3 = (df1["Year"] == 2020) df4 = df1.loc[(cond1 & cond2) | cond3]`
- 증감률이 0보다 작다는 것은 매출액이 줄었다는 것을 뜻한다.

- Mission
> '01_Contract_Data.csv' 데이터의 'Datetime'항목을 처리하여, 요일 데이터를 추출한 뒤, 요일과 제품에 따른 'Amount_Month'의 합을 계산 해 보세요!

In [63]:
# 17. 19년도 1-8월, 20년도 전체(1-8월) product type에 따른 연도별 Amount_Month의 sum
df4.pivot_table(index="Product_Type", columns="Year", values="Amount_Month", aggfunc="sum")

Year                2019       2020
Product_Type                       
DES-1         2236446810  748837116
DES-2          486777810  155779320
DES-3A         117017160   39922500
DES-R4          25721700    8437200
ERA             46972800   14171100
MMC             14077500    5193900

In [66]:
# 18. 1-8월 증감액 & 증감률
df5 = df4.pivot_table(index="Product_Type", columns="Year", values="Amount_Month", aggfunc="sum")
df5["증감액"] = df5[2020] - df5[2019]
df5["증감률"] = df5["증감액"] / (df5[2020] + df5[2019])
df5

Year                2019       2020         증감액       증감률
Product_Type                                             
DES-1         2236446810  748837116 -1487609694 -0.498314
DES-2          486777810  155779320  -330998490 -0.515127
DES-3A         117017160   39922500   -77094660 -0.491238
DES-R4          25721700    8437200   -17284500 -0.506003
ERA             46972800   14171100   -32801700 -0.536467
MMC             14077500    5193900    -8883600 -0.460973

In [75]:
# Mission
df = pd.read_csv("dataset/01_Contract_Data.csv")
df["Datetime(dt)"] = pd.to_datetime(df["Datetime"], format="%Y-%m-%d")
df["요일"] = df["Datetime(dt)"].dt.day_name()
df.pivot_table(index=["요일", "Product_Type"], values="Amount_Month", aggfunc="sum")

Amount_Month
요일        Product_Type              
Friday    DES-1            542749254
          DES-2            116072160
          DES-3A            28468500
          DES-R4             7143600
          ERA               13280400
          MMC                3402600
Monday    DES-1            540517839
          DES-2            112428060
          DES-3A            28965600
          DES-R4             4967400
          ERA                9247800
          MMC                2659800
Saturday  DES-1            406091973
          DES-2             87389670
          DES-3A            21167550
          DES-R4             4903800
          ERA                9570000
          MMC                3216900
Sunday    DES-1            388083615
          DES-2             79818960
          DES-3A            20073150
          DES-R4             4060800
          ERA                8028300
          MMC                1917000
Thursday  DES-1            615978612
          DES-2            136978050
          DES-3A            34292460
          DES-R4             7416900
          ERA               10898400
          MMC                4762500
Tuesday   DES-1            607328556
          DES-2            130079850
          DES-3A            30982350
          DES-R4             6651300
          ERA               14400000
          MMC                3156900
Wednesday DES-1            594365877
          DES-2            124890540
          DES-3A            33135450
          DES-R4             7184700
          ERA               11690700
          MMC                3618300

In [74]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51301 entries, 0 to 51300
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Index          51301 non-null  int64         
 1   Member_ID      51301 non-null  int64         
 2   Sales_Type     51301 non-null  object        
 3   Contract_Type  51301 non-null  object        
 4   Channel        51301 non-null  object        
 5   Datetime       51301 non-null  object        
 6   Term           51301 non-null  int64         
 7   Payment_Type   51301 non-null  object        
 8   Product_Type   51301 non-null  object        
 9   Amount_Month   51301 non-null  int64         
 10  Customer_Type  51299 non-null  object        
 11  Age            44329 non-null  float64       
 12  Address1       51299 non-null  object        
 13  Address2       51299 non-null  object        
 14  State          51301 non-null  object        
 15  Overdue_count  5130